Adds simulation and mockup threads for drilling machine plugin

moriarty opened the pull request:
Adds drilling_machine_simulation_thread and drilling_machine_mockup_thread

In simulation mode, the drilling machine the operation is simulated.
In mockup mode, clips functions are exposed but nothing is simulated.

This pull request closes #10. Feedback and re-factoring has been incorporated.

moriarty committed:
plugins: add simulation of the drilling machine

- Adds simulation and mockup threads for drilling machine plugin
- Default Drilling Machine config to simulation

svenschneider commented on cfg/config.yaml:
```diff
@@ -18,7 +18,7 @@ llsfrb:
 
   drilling-machine:
-    enable: false
+    mode: mockup
```
In my opinion we can enable the simulation by default. What do you think?

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,145 @@
+    fawkes::MutexLocker lock2(&drill_machine_mutex_);
+
+    // First time
+    if (!last_status_msg_.has_state())
```
Why is this conditional? It's probably a good idea to always initialize the state (in the init function).

moriarty replied:
Will fix

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,145 @@
+{
+}
+
+void DrillingMachineSimulationThread::loop()
```
Shouldn't the mutex be locked here as well (i.e. each time you access the command and state)? Else it is somehow pointless: init and loop will never be executed concurrently, but the loop and each CLIPS access is concurrent.

moriarty replied:
The lock is now two lines lower- so github isn't considering this a comment "on an outdated diff".

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,145 @@
+
+DrillingMachineStatus::State DrillingMachineSimulationThread::clips_get_device_state()
+{
+    if (last_status_msg_.has_state())
```
Lock mutex (see above)?

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,145 @@
+
+int DrillingMachineSimulationThread::clips_is_device_connected()
+{
+    return (last_status_msg_.has_is_device_connected() && last_status_msg_.is_device_connected());
```
Lock mutex (see above)?

svenschneider commented:
Besides my comments this looks good. Please fix or provide feedback.

moriarty commented:
@svenschneider - feedback has been integrated

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,144 @@
+
+void DrillingMachineSimulationThread::loop()
+{
+    fawkes::MutexLocker lock2(&drill_machine_mutex_);
```
This together with the last line in this function leads to a horrible lockup of the complete CFH:
- The _scoped_ lock is acquired here
- In the last line the thread goes to sleep, but does not release the lock. Therefore, all calls that also need to lock the mutex will wait.

My proposal is to manually release the lock before sending the thread to sleep.

Additionally, why is the variable here (and in other places) called lock2?

svenschneider commented:
Please integrate the feedback, squash the commits appropriately and update the pull request.

When squashing the commits please also update the commit messages. The convention for the subject line is "component: message", where the message is all lower-case and limited to 50 characters. Component is the "top-level" folder of the changed file(s), here component should be "plugins" (within the rockin folder).

svenschneider commented:
Any update?

moriarty commented:
@svenschneider - update, I fixed all the feedback and rebased squashing commits. 

I used

```
git push -f moriarty drill-plugin-sim-rebased:drill-plugin-sim
```

There are still some comments about locking the mutex. Should I be locking it there?
The value is being read. I don't think we're locking it their in the other device threads.

svenschneider commented:
The existing code (hopefully) may work because the variable that is read, and concurrently written from somewhere else, atomically. Therefore, the assumptions for this code being correct are:
1. This variable is accessible atomically on all hardware platforms (we want to use).
2. ProtoBuf always maps the variable to the "right" data type.
If the assumptions don't hold, the reading may result in corrupt data.

Thus, in my opinion the safe way is to lock the mutex for each access (read and write). Could you please check if this has any noticable effect?

moriarty committed:
plugins: drilling machine lock(clips_mutex) when reading

moriarty committed:
plugins: drilling machine simulation lock(clips_mutex)

moriarty commented:
@svenschneider I've added the lock(clips_mutex) to both the simulation and real thread. 
I did both on this branch.

svenschneider commented on rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp:
```diff
@@ -0,0 +1,142 @@
+    last_command_msg_.set_command(drill_command);
+    last_sent_command_timestamp_ = boost::posix_time::microsec_clock::local_time();
+
+    logger->log_info("DrillingMachineSim", "Send drill command: %d", drill_command);
```
I just saw that this is "spamming" the terminal. Therefore, I would suggest to use "logger->log_debug([...])".

moriarty replied:
should I modify that in the other plugins as well? And the non-simulated versions?

svenschneider replied:
Yes, would be nice.

svenschneider commented:
Thanks, there is one more minor issue about the logging. When that feedback is integrated, we should be ready to go.

moriarty committed:
plugins: drilling machine clean up debug output

moriarty commented:
I removed some of the logging and switched some log_debug as suggested.
